### ISSUE TITLE
Unhandled exception in parser

### ISSUE THREAD

**Issue by usernameiwantedwasalreadytaken** [ToruNiina/toml11#8]
Hi,

I've noticed that inside the `parser.hpp` you throw some `std::pair`s as exceptions, then you catch them inside the parse function and rethrow only the syntax_error part.

That's fine, but in one of my tests, I used an invalid toml file (with two "[table]", instead of "[[table]]") and I've got an unhandled exception.

The code inside the parse function is like this (lines 1081-1088):
```cpp
     // ...
    typedef std::vector<toml::character>::const_iterator iterator_type;
    is.read(contents.data(), size);
    try
    {
        return parse_data::invoke(contents.cbegin(), contents.cend());
    }
    catch(std::pair<iterator_type, syntax_error> iter_except)
    {
     // ...
```

But in this specific case, the exception being thrown is not caught by that. I managed to catch it with code like this:

```cpp
  toml::Table data;
  typedef std::vector<std::string>::iterator iterator_type;
  try {
    data = parse(path.string());
  } catch(const std::pair<iterator_type, syntax_error>& e) {
    fatal(e.second.what());
  }
```

Note the different iterator_type.

Maybe there is something missing in the parse method...

**Comment by ToruNiina** [comment]
Hi,

I found the exception that make the problem in the method that pushes new table into array_of_table.
There, iterator become invalid because it points to automatic variable.

Since the syntax_error thrown from the function already has the information that the iterator has, I removed iterator from the exception type.

Now (5aeb6f2), only syntax_error will be thrown when key is duplicated.
Thank you for pointing the problem out.

Best,
Niina
